lib/zephyr/src/zephyr/backends.py

```diff
@@ -640,241 +163,298 @@ def execute(self, dataset: Dataset[T], hints: ExecutionHint = ExecutionHint(), v
+                execution_context=self.context,
+            )
             for shard_idx, (shard, aux_shards) in enumerate(zip(shards, aux_shards_per_shard, strict=True))
         ]
 
-        chunks_by_shard_idx = defaultdict(list)
-
-        if shard_tasks:
-            active_gens = []
-            queued_tasks = shard_tasks[:]
-            finished_gens = []
-
-            while len(active_gens) < self.config.max_parallelism and queued_tasks:
-                shard_idx, shard, aux_shards = queued_tasks.pop(0)
-                ctx = ApplyShardCtx(
-                    shard=shard,
-                    shard_idx=shard_idx,
-                    total_shards=total,
-                    chunk_size=hints.chunk_size,
-                    aux_shards=aux_shards,
-                )
-                active_gens.append(self.context.run(process_fn, ctx, *fn_args))
-
-            while active_gens or queued_tasks:
-                ready, _ = self.context.wait(active_gens, num_returns=1)
-
-                for ready_gen in ready:
-                    try:
-                        header = self.context.get(next(ready_gen))
-                        data = next(ready_gen)  # ref to header data
-                        chunks_by_shard_idx[header.shard_idx].append(Chunk(header.count, data))
-
-                    except StopIteration:
-                        finished_gens.append(ready_gen)
-                        active_gens.remove(ready_gen)
-                        if queued_tasks:
-                            shard_idx, shard, aux_shards = queued_tasks.pop(0)
-                            ctx = ApplyShardCtx(
-                                shard=shard,
-                                shard_idx=shard_idx,
-                                total_shards=total,
-                                chunk_size=hints.chunk_size,
-                                aux_shards=aux_shards,
-                            )
-                            active_gens.append(self.context.run(process_fn, ctx, *fn_args))
-
-        # Reconstruct shards from collected chunks
-        if len(chunks_by_shard_idx) == 0:
-            return []
+        # Run tasks
+        results = self._run_tasks(contexts, operations)
+
+        # Build Shards from results
+        # Use input shard count to preserve empty shards (important for joins)
+        num_output_shards = total
+        if results:
+            # If scatter/groupby produces more shards, use that count
+            max_result_idx = max(results.keys())
+            if max_result_idx >= num_output_shards:
+                num_output_shards = max_result_idx + 1
 
-        max_shard_idx = max(chunks_by_shard_idx.keys())
         return [
-            Shard(idx=idx, chunks=chunks_by_shard_idx.get(idx, []), context=self.context)
-            for idx in range(max_shard_idx + 1)
+            Shard(
+                idx=idx,
+                chunks=[Chunk(header.count, data_ref) for _, header, data_ref in results.get(idx, [])],
```

Comment:
Is the default `[]` correct here? afaiu if it was used, it would fail to unpack the 3 values, no?

Reply:
it looks like i had a stale push. the code structure is confusing. the [] just means we end up a with a shard with no chunks in it, we shouldn't try to unpack anything.

i adjusted the structure to make this more obvious now